gnovm/pkg/gnolang/preprocess.go

```diff
@@ -2333,11 +2340,50 @@ func checkOrConvertType(store Store, last BlockNode, x *Expr, t Type, autoNative
 					// default:
 				}
 			}
+			// convert x to destination type t
 			cx := Expr(Call(constType(nil, t), *x))
 			cx = Preprocess(store, last, cx).(Expr)
 			*x = cx
+		} else {
+			// if one side is declared name type and the other side is unnamed type
+			if isNamedConversion(xt, t) {
+				// covert right (xt) to the type of the left (t)
+				cx := Expr(Call(constType(nil, t), *x))
+				cx = Preprocess(store, last, cx).(Expr)
+				*x = cx
+			}
+		}
+	}
+}
+
+// Return true if we need to convert named and unnamed types in an assignment
+func isNamedConversion(xt, t Type) bool {
+	if t == nil {
+		t = xt
+	}
+
+	// t is left hand destination type, xt is right hand expression type
+	// In a few special cases, we should not consider compare named and unnamed type
+	// case 1: if left is interface, which is unnamed, we dont convert to the left type even right is named type.
+
+	_, c1 := t.(*InterfaceType)
+
+	// case2: TypeType is used in make() new() native uverse definition and TypeType.IsNamed() will panic on unexpected.
+
+	_, oktt := t.(*TypeType)
+	_, oktt2 := xt.(*TypeType)
+	c2 := oktt || oktt2
+
+	//
+	if !c1 && !c2 { // carve out above two cases
+		// covert right to the type of left if one side is unnamed type and the other side is not
+
+		if t.IsNamed() && !xt.IsNamed() ||
+			!t.IsNamed() && xt.IsNamed() {
+			return true
 		}
 	}
+	return false
```

Comment:
I understand a layer of abstraction makes the code looks tighter. However, we want to be explicit why we implement this logic. 

 ` return t.IsNamed() != xt.IsNamed()`  look short and have the same logic blow.

However, I still prefer following code since it literally translate the specs and the comments. 

            // normal case: convert for named = unnamed, or unnamed = named assignments.

	    // covert right to the type of left if one side is unnamed type and the other side is not

		    if t.IsNamed() && !xt.IsNamed() ||
			    !t.IsNamed() && xt.IsNamed() {
			    return true
		    }

Reply:
I'm fine if you prefer `t.IsNamed() && !xt.IsNamed() || !t.IsNamed() && xt.IsNamed()`. The `!=` is a neat logic trick which to my interpretation of code is more readable; I understand wanting to write the conditions explicitly though, and that all of our intelligences work differently, so that's fine.

I do still think that we should do this:

```go
return t.IsNamed() && !xt.IsNamed() ||
	!t.IsNamed() && xt.IsNamed()
```

An `if` on that which returns true, followed by a return false, to me just adds visual clutter and harms readability.

Reply:
Thank you for the suggestions. I prefer explicitly returning 'true' or 'false' for boolean returns, there will no second guess on the intention of the logic.